File: factorcombinations.js
//Objective is to find all unique factor combinations for a certain number.
//Factors must not include 1.

let num = 8


//O() backtracking solution.

let temp = []
let result = []
let index = 2
let product = 1

backtrack(temp, product, index)

function backtrack(temp, product, index) {
    if (product > num) {
        return
    }

    if (product == num) {
        result.push([...temp])
        return
    }

    //Here we only have to check up to (num / product) since every time
    //we multiply the product by an integer, the values we can iterate
    //through to find a match decreases.
    //Make sure i !== num to avoid (1, 8)
    for (let i = index; i <= Math.floor(num / product); i++) {
        if (num % i == 0 && i !== num) {
            temp.push(i)
            backtrack(temp, product * i, i)
            temp.pop()
        }
    }
}

return result

File: factorcombinations.java
//Java Solution

class Solution {
    public List<List<Integer>> getFactors(int n) {
        if (n <= 1) {
            return new ArrayList<>();
        }
        
        List<List<Integer>> result = new ArrayList<>();
        LinkedList<Integer> temp = new LinkedList<>();
        int product = 1;
        int index = 2;
        
        backtrack(temp, result, product, index, n);
        return result;
    }
    
    public void backtrack(LinkedList<Integer> temp, List<List<Integer>> result, int product, int index, int n) {
        if (product > n) {
            return;
        }
        
        if (product == n) {
            result.add(new ArrayList<>(temp));
            return;
        }
        
        for (int i = index; i <= (n / product); i++) {
            if (n % i == 0 && n != i) {
                temp.add(i);
                backtrack(temp, result, product * i, i, n);
                temp.removeLast();
            }
        }
    }
}